numpy/lib/arraypad.py

```diff
@@ -69,821 +82,484 @@ def _round_ifneeded(arr, dtype):
         Axis along which to pad `arr`.
+    index_pair : tuple of ints, length 2
+        Padding to (prepend, append) along `axis`.
 
     Returns
     -------
-    padarr : ndarray
-        Output array, with `pad_amt[0]` values prepended and `pad_amt[1]`
-        values appended along `axis`. Both regions are padded wrapped values
-        from the opposite end of `axis`.
+    pad_amt : tuple of ints, length 2
+        New index positions of padding to do along the `axis`. If these are
+        both 0, padding is done in this dimension. See notes on why this is
+        necessary.
 
     Notes
     -----
-    This method of padding is also known as 'tile' or 'tiling'.
+    This algorithm does not pad with repetition, i.e. the edges are not
+    repeated in the reflection. For that behavior, use `mode='symmetric'`.
 
     The modes 'reflect', 'symmetric', and 'wrap' must be padded with a
     single function, lest the indexing tricks in non-integer multiples of the
     original shape would violate repetition in the final iteration.
-
     """
-    # Implicit booleanness to test for zero (or None) in any scalar type
-    if pad_amt[0] == 0 and pad_amt[1] == 0:
-        return arr
-
-    ##########################################################################
-    # Prepended region
-
-    # Slice off a reverse indexed chunk from near edge to pad `arr` before
-    wrap_slice = _slice_last(arr.shape, pad_amt[0], axis=axis)
-    wrap_chunk1 = arr[wrap_slice]
-
-    ##########################################################################
-    # Appended region
-
-    # Slice off a reverse indexed chunk from far edge to pad `arr` after
-    wrap_slice = _slice_first(arr.shape, pad_amt[1], axis=axis)
-    wrap_chunk2 = arr[wrap_slice]
-
-    # Concatenate `arr` with both chunks, extending along `axis`
-    return np.concatenate((wrap_chunk1, arr, wrap_chunk2), axis=axis)
+    left_pad, right_pad = index_pair
+    period = padded.shape[axis] - right_pad - left_pad
+
+    # If the current dimension of `arr` doesn't contain enough valid values
+    # (not part of the undefined pad area) we need to pad multiple times.
+    # Each time the pad area shrinks on both sides which is communicated with
+    # these variables.
+    new_left_pad = 0
+    new_right_pad = 0
+
+    if left_pad > 0:
+        # Pad with wrapped values on left side
+        # First slice chunk from right side of the non-pad area.
+        # Use min(period, left_pad) to ensure that chunk is not larger than
+        # pad area
+        right_slice = _slice_at_axis(
+            padded.shape,
+            slice(-right_pad - min(period, left_pad),
+                  -right_pad if right_pad != 0 else None),
+            axis,
+        )
+        right_chunk = padded[right_slice]
+
+        if left_pad > period:
+            # Chunk is smaller than pad area
+            pad_area = _slice_at_axis(
+                padded.shape, slice(left_pad - period, left_pad), axis
+            )
```

Comment:
I prefer the bracket on the last line with functions.

Reply:
[PEP 8](https://www.python.org/dev/peps/pep-0008/#indentation) recommends a few options. I would prefer a comma after `axis`, or maybe even
```
pad_area = _slice_at_axis(
                padded.shape, 
                slice(left_pad - period, left_pad), 
                axis,
            )
```
Definitely the double indent is preferable

Reply:
It looks like either @lagru's way or @seberg's way is blessed by PEP8.

Using multiple levels of indentation to align by column tends to be a bad idea, as things get out of alignment when functions get renamed - it happens all the time in our C code.

Reply:
I am fine with both, somehow used to the other for functions I guess.

Reply:
I wish more projects used [black](https://github.com/ambv/black) which would helps saving review time on such code style discussions. Though it does seem difficult to adapt for large projects. (FWIW here it would format that line as it was initially).

Reply:
For what it's worth I generally had good experiences with black. But for nested structures formatting by hand is sometimes more readable and black would be one more developer tool for contributors to learn / deal with. It will almost always produce readable and consistent formatting though. And removes the "dread" of going over the line. :D